services/src/main/java/org/keycloak/authentication/requiredactions/UpdateTotp.java

```diff
@@ -69,47 +91,82 @@ public void requiredActionChallenge(RequiredActionContext context) {
     public void processAction(RequiredActionContext context) {
         EventBuilder event = context.getEvent();
         event.event(EventType.UPDATE_TOTP);
+
+        Response challenge;
+        if (context.getAuthenticationSession() == null) {
```

Comment:
Same comment like above. I think this null check is not needed here (the case with invalid/expired authenticationSession is handled earlier in the request). So the whole snippet of lines 96-101 an be hopefully removed?

Reply:
removed this check and extracted to a method